File: LinkedList/PalindromeLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
    ListNode* reverse(ListNode* head){
        ListNode* prev = NULL;
        ListNode* ne = NULL;
        while(head){
            ne = head->next;
            head->next = prev;
            prev = head;
            head = ne;
        }
        return prev;
    }
public:
    bool isPalindrome(ListNode* head) {
        if(!head||head->next==NULL)
            return true;
        
        ListNode* slow = head;
        ListNode* fast = head;
        
        while(fast->next && fast->next->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        
        //reverse the linked list after the slow pointer
        slow->next = reverse(slow->next);
        
        //move the slow to the start of the reversed part of LL
        slow = slow->next;
        
        while(slow){
            if(slow->val != head->val)
                return false;
            slow = slow->next;
            head = head->next;
        }
        
        return true;
    }
};

File: palindromeLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    //tc - O(N/2)+O(N/2)+O(N/2) sc - O(1)
    
    //function to reverse a linked list
    ListNode reverse(ListNode head){
        ListNode prev = null;
        while(head!=null){
            ListNode next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }
        return prev;
    }
    
    public boolean isPalindrome(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        ListNode start = head;
        while(fast.next!=null && fast.next.next!=null){
            slow = slow.next;
            fast = fast.next.next;
        }
        
        //reverse the second half of the list
        slow.next = reverse(slow.next);
        //move slow to point to first element in the second half
        slow = slow.next;
        
        //take one pointer to the first element and move along the second half of the list
        while(slow!=null){
            if(start.val!=slow.val)
                return false;
            slow = slow.next;
            start = start.next;
        }
        
        return true;
    }
}